### PR TITLE
Add upsert search attributes

### PR DESCRIPTION
**Pull request by vancexu**

### CODE REVIEW

File: internal/internal_workflow_testsuite.go [hunk 1]
```diff
@@ -181,6 +181,8 @@ type (
 
 		workerStopChannel  chan struct{}
 		sessionEnvironment *testSessionEnvironmentImpl
+
+		searchAttributes map[string]interface{}
```
Comment (yycptt) [inline internal/internal_workflow_testsuite.go]: Is it used somewhere?

File: internal/internal_event_handlers.go [hunk 2]
```diff
@@ -288,6 +288,29 @@ func (wc *workflowEnvironmentImpl) SignalExternalWorkflow(domainName, workflowID
...
+	if err != nil {
+		return err
+	}
+
+	upsertID := wc.GenerateSequenceID()
+	wc.decisionsHelper.upsertSearchAttributes(upsertID, attr)
+	return nil
+}
+
+// This has to be used in workflowEnvironment implementations instead of in Workflow for testsuite mock purpose.
```
Comment (yycptt) [inline internal/internal_event_handlers.go]: I think this comment should be moved to L292.

File: internal/workflow.go [hunk 3]
```diff
@@ -805,6 +806,43 @@ func signalExternalWorkflow(ctx Context, workflowID, runID, signalName string, a
...
+//   	"CustomIntField": 2,
+//   	"CustomBoolField": true,
+//   	"CustomKeywordField": "seattle",
+//   }
+// This is only supported when using ElasticSearch.
+func UpsertSearchAttributes(ctx Context, attributes map[string]interface{}) error {
+	ctx1 := setWorkflowEnvOptionsIfNotExist(ctx)
+	options := getWorkflowEnvOptions(ctx1)
+
+	if options.domain == nil || *options.domain == "" {
```
Comment (yycptt) [inline internal/workflow.go]: I don't quite understand why we need this check?

Comment (vancexu) [reply internal/workflow.go]: Good catch, I copied from other method but found no actually needed

File: internal/internal_workflow_testsuite.go [hunk 4]
```diff
@@ -227,6 +227,8 @@ func newTestWorkflowEnvironmentImpl(s *WorkflowTestSuite) *testWorkflowEnvironme
 
 		doneChannel:       make(chan struct{}),
 		workerStopChannel: make(chan struct{}),
+
+		searchAttributes: make(map[string]interface{}),
```
Comment (yycptt) [inline internal/internal_workflow_testsuite.go]: Remove this too.

**Pull request merged**